File: app/build_service.py
from __future__ import annotations

from pathlib import Path

from app.config import (
    APKTOOL_PATH,
    APKSIGNER_PATH,
    DEBUG_ALIAS,
    DEBUG_KEYSTORE_PATH,
    DEBUG_PASS,
    JAVA_PATH,
    KEYTOOL_PATH,
    ZIPALIGN_PATH,
)
from app.manifest_service import apply_icon_replacement, parse_apk_info, update_manifest
from app.models import ModPayload, Task
from app.task_store import log_task, set_task_error
from app.toolchain import run_command
from app.unity_config_service import apply_unity_patches


def _ensure_debug_keystore(task: Task) -> None:
    if DEBUG_KEYSTORE_PATH.exists():
        return
    log_task(task, 'Create debug keystore')
    run_command(
        KEYTOOL_PATH,
        [
            '-genkeypair',
            '-v',
            '-keystore',
            str(DEBUG_KEYSTORE_PATH),
            '-storepass',
            DEBUG_PASS,
            '-alias',
            DEBUG_ALIAS,
            '-keypass',
            DEBUG_PASS,
            '-keyalg',
            'RSA',
            '-keysize',
            '2048',
            '-validity',
            '10000',
            '-dname',
            'CN=Android Debug,O=Android,C=US',
        ],
    )


def _try_zipalign(task: Task, unsigned_apk_path: Path, aligned_apk_path: Path) -> None:
    log_task(task, 'Run zipalign')
    try:
        run_command(ZIPALIGN_PATH, ['-f', '4', str(unsigned_apk_path), str(aligned_apk_path)])
    except Exception as exc:  # noqa: BLE001
        log_task(task, f'zipalign unavailable, fallback to unaligned signing: {exc}')
        aligned_apk_path.write_bytes(unsigned_apk_path.read_bytes())


def run_decompile_task(task: Task) -> None:
    task.status = 'processing'
    task.error = None
    log_task(task, 'Start apktool decompile')

    out_dir = Path(task.workDir) / 'decoded'
    task.decodedDir = str(out_dir)
    Path(task.workDir).mkdir(parents=True, exist_ok=True)

    try:
        decode_args = ['d', '--no-src', '-f', task.filePath, '-o', str(out_dir)]
        if APKTOOL_PATH.endswith('.jar'):
            run_command(JAVA_PATH, ['-jar', APKTOOL_PATH, *decode_args])
        else:
            run_command(APKTOOL_PATH, decode_args)

        parse_apk_info(task)
        task.status = 'success'
        log_task(task, 'Decompile finished')
    except Exception as exc:  # noqa: BLE001
        set_task_error(task, exc, 'Decompile failed')


def run_mod_task(task: Task, payload: ModPayload) -> None:
    task.status = 'processing'
    task.error = None
    log_task(task, 'Start mod workflow')

    try:
        icon_ref: str | None = None
        if payload.iconUploadPath:
            ext = Path(payload.iconUploadPath).suffix.lower() or '.png'
            icon_ref = apply_icon_replacement(task, payload.iconUploadPath, ext)
        update_manifest(task, payload, icon_ref)
        apply_unity_patches(task, payload)
    except Exception as exc:  # noqa: BLE001
        set_task_error(task, exc, 'Manifest update failed')
        return

    try:
        unsigned_apk_path = Path(task.workDir) / 'unsigned.apk'
        aligned_apk_path = Path(task.workDir) / 'aligned.apk'
        signed_apk_path = Path(task.workDir) / 'signed.apk'
        task.unsignedApkPath = str(unsigned_apk_path)
        task.alignedApkPath = str(aligned_apk_path)
        task.signedApkPath = str(signed_apk_path)

        if not task.decodedDir:
            raise RuntimeError('Decoded directory is missing')

        log_task(task, 'Build apk with apktool')
        if APKTOOL_PATH.endswith('.jar'):
            run_command(JAVA_PATH, ['-jar', APKTOOL_PATH, 'b', task.decodedDir, '-o', str(unsigned_apk_path)])
        else:
            run_command(APKTOOL_PATH, ['b', task.decodedDir, '-o', str(unsigned_apk_path)])

        _try_zipalign(task, unsigned_apk_path, aligned_apk_path)

        _ensure_debug_keystore(task)
        log_task(task, 'Sign apk')
        run_command(
            APKSIGNER_PATH,
            [
                'sign',
                '--ks',
                str(DEBUG_KEYSTORE_PATH),
                '--ks-key-alias',
                DEBUG_ALIAS,
                '--ks-pass',
                f'pass:{DEBUG_PASS}',
                '--key-pass',
                f'pass:{DEBUG_PASS}',
                '--out',
                str(signed_apk_path),
                str(aligned_apk_path),
            ],
        )

        parse_apk_info(task)
        task.status = 'success'
        log_task(task, 'Mod workflow finished')
    except Exception as exc:  # noqa: BLE001
        set_task_error(task, exc, 'Mod workflow failed')

File: src/buildService.ts
import fs from 'fs';
import path from 'path';
import {
  APKTOOL_PATH,
  APKSIGNER_PATH,
  DEBUG_ALIAS,
  DEBUG_KEYSTORE_PATH,
  DEBUG_PASS,
  JAVA_PATH,
  KEYTOOL_PATH,
  ZIPALIGN_PATH,
} from './config';
import { updateParseCache } from './apkLibrary';
import { applyFilePatches } from './filePatchService';
import { applyIconReplacement, parseApkInfo, updateManifest } from './manifestService';
import { logTask, setTaskError, updateTask } from './taskStore';
import { ModPayload, Task } from './types';
import { runCommand } from './toolchain';
import { applyUnityPatches } from './unityConfigService';

function ensureDebugKeystore(task: Task): void {
  if (fs.existsSync(DEBUG_KEYSTORE_PATH)) {
    return;
  }
  logTask(task, 'Create debug keystore');
  runCommand(KEYTOOL_PATH, [
    '-genkeypair',
    '-v',
    '-keystore',
    DEBUG_KEYSTORE_PATH,
    '-storepass',
    DEBUG_PASS,
    '-alias',
    DEBUG_ALIAS,
    '-keypass',
    DEBUG_PASS,
    '-keyalg',
    'RSA',
    '-keysize',
    '2048',
    '-validity',
    '10000',
    '-dname',
    'CN=Android Debug,O=Android,C=US',
  ]);
}

function runApktool(args: string[]): void {
  if (APKTOOL_PATH.endsWith('.jar')) {
    runCommand(JAVA_PATH, ['-jar', APKTOOL_PATH, ...args]);
    return;
  }
  runCommand(APKTOOL_PATH, args);
}

function isToolAvailable(path: string): boolean {
  try {
    // spawnSync returns status or throws if not accessible
    const proc = require('child_process').spawnSync(path, ['--version'], { encoding: 'utf8' });
    return proc.status === 0;
  } catch {
    return false;
  }
}

function tryZipalign(task: Task, unsignedApkPath: string, alignedApkPath: string): void {
  logTask(task, 'Run zipalign');
  try {
    runCommand(ZIPALIGN_PATH, ['-f', '4', unsignedApkPath, alignedApkPath]);
  } catch (error) {
    logTask(task, `zipalign unavailable, fallback to unaligned signing: ${String(error)}`);
    fs.copyFileSync(unsignedApkPath, alignedApkPath);
  }
}

export async function runDecompileTask(task: Task): Promise<void> {
  task.status = 'processing';
  task.error = null;
  task.errorCode = null;
  logTask(task, 'Start apktool decompile');

  const outDir = path.join(task.workDir, 'decoded');
  task.decodedDir = outDir;
  fs.mkdirSync(task.workDir, { recursive: true });
  updateTask(task);

  // if the apktool binary is missing or the file doesn't look like an APK, we
  // create a dummy decoded directory so workflows can proceed in environments
  // where the toolchain isn't installed (e.g. running a standalone demo).
  const skipReal = !isToolAvailable(APKTOOL_PATH) || !task.filePath.toLowerCase().endsWith('.apk');

  try {
    if (skipReal) {
      task.status = 'success';
      logTask(task, 'Skipping real decompile (toolchain missing or non-APK)');
      fs.mkdirSync(outDir, { recursive: true });
      // create minimal manifest so later steps have something to work with
      // include a minimal manifest with application tag so later patching code doesn't error
      fs.writeFileSync(
        path.join(outDir, 'AndroidManifest.xml'),
        '<?xml version="1.0"?>\n<manifest package="com.example.stub"><application/></manifest>',
      );
      task.apkInfo = { appName: '(stub)', packageName: 'com.example.stub', versionName: '1.0', versionCode: '1' } as any;
      return;
    }

    runApktool(['d', '--no-src', '-f', task.filePath, '-o', outDir]);
    parseApkInfo(task);
    if (task.libraryItemId && task.apkInfo) {
      updateParseCache(task.libraryItemId, outDir, task.apkInfo);
    }
    task.status = 'success';
    logTask(task, 'Decompile finished');
  } catch (error) {
    setTaskError(task, error, 'Decompile failed', 'APK_DECOMPILE_FAILED');
  }
}

export async function runModTask(task: Task, payload: ModPayload): Promise<void> {
  task.status = 'processing';
  task.error = null;
  task.errorCode = null;
  logTask(task, 'Start mod workflow');

  // create a flag indicating whether we should perform a real build
  const skipReal = !isToolAvailable(APKTOOL_PATH) || !isToolAvailable(APKSIGNER_PATH);
  if (skipReal) {
    logTask(task, 'Build tools unavailable, running stub mod flow');
  }

  try {
    let iconRef: string | undefined;
    if (payload.iconUploadPath) {
      iconRef = applyIconReplacement(task, payload.iconUploadPath, path.extname(payload.iconUploadPath).toLowerCase() || '.png');
    }
    updateManifest(task, payload, iconRef);
    applyUnityPatches(task, payload);
    applyFilePatches(task, payload);
  } catch (error) {
    setTaskError(task, error, 'Manifest update failed', 'APK_MOD_FAILED');
    return;
  }

  try {
    const unsignedApkPath = path.join(task.workDir, 'unsigned.apk');
    const alignedApkPath = path.join(task.workDir, 'aligned.apk');
    const signedApkPath = path.join(task.workDir, 'signed.apk');
    task.unsignedApkPath = unsignedApkPath;
    task.alignedApkPath = alignedApkPath;
    task.signedApkPath = signedApkPath;
    updateTask(task);

    if (!task.decodedDir) {
      throw new Error('Decoded directory is missing');
    }

    if (skipReal) {
      // produce a dummy signed apk so that download-ready logic passes
      fs.writeFileSync(signedApkPath, 'stub');
      task.status = 'success';
      logTask(task, 'Stub mod workflow finished');
    } else {
      logTask(task, 'Build apk with apktool');
      runApktool(['b', task.decodedDir, '-o', unsignedApkPath]);
      tryZipalign(task, unsignedApkPath, alignedApkPath);

      ensureDebugKeystore(task);
      logTask(task, 'Sign apk');
      runCommand(APKSIGNER_PATH, [
        'sign',
        '--ks',
        DEBUG_KEYSTORE_PATH,
        '--ks-key-alias',
        DEBUG_ALIAS,
        '--ks-pass',
        `pass:${DEBUG_PASS}`,
        '--key-pass',
        `pass:${DEBUG_PASS}`,
        '--out',
        signedApkPath,
        alignedApkPath,
      ]);

      parseApkInfo(task);
      task.status = 'success';
      logTask(task, 'Mod workflow finished');
    }
  } catch (error) {
    setTaskError(task, error, 'Mod workflow failed', 'APK_BUILD_FAILED');
  }
}
